### PR TITLE
Jb cp nc user voting

### PR DESCRIPTION
**Pull request by chanhpark**

### CODE REVIEW

File: app/controllers/votes_controller.rb [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+    # Instead of update method
+    begin
+      if @vote.save
+        redirect_to site_path(@site)
+        flash[:alert] = "Vote has been recorded"
+      else
+        redirect_to site_path(@site)
+        flash[:alert] = "Vote Already Recorded"
+      end
+    rescue
```
Comment (dpickett) [inline app/controllers/votes_controller.rb]: Why is exception handling being utilized here? 

http://launchware.com/articles/catch-all-exception-handling-is-not-flow-control

File: app/models/review.rb [hunk 2]
```diff
@@ -1,4 +1,11 @@
 class Review < ActiveRecord::Base
   belongs_to :site
   belongs_to :user
+
+  has_many :votes
+  accepts_nested_attributes_for :votes
+
+  def total_score
```
Comment (dpickett) [inline app/models/review.rb]: excellent! This is a function and concern of the review.

File: app/models/vote.rb [hunk 3]
```diff
@@ -0,0 +1,6 @@
+class Vote < ActiveRecord::Base
+  belongs_to :review
+  belongs_to :user
+
+  validates_uniqueness_of :choice, scope: [:user_id, :review_id]
```
Comment (dpickett) [inline app/models/vote.rb]: great validation. We also require presence of a choice, too, right?

File: config/routes.rb [hunk 4]
```diff
@@ -1,6 +1,10 @@
 Rails.application.routes.draw do
   resources :sites do
-    resources :reviews
+    resources :reviews, only: [:index, :new, :create, :edit, :update, :destroy]
```
Comment (dpickett) [inline config/routes.rb]: excellent configuration of your routes. this is really the ideal and most adequate organization of concerns

File: db/migrate/20150106183347_create_votes_table.rb [hunk 5]
```diff
@@ -0,0 +1,11 @@
+class CreateVotesTable < ActiveRecord::Migration
+  def change
+    create_table :votes do |t|
+      t.boolean :up_vote, null: false, default: false
```
Comment (dpickett) [inline db/migrate/20150106183347_create_votes_table.rb]: should we consolidate the vote to a single field for easier aggregation?

File: db/migrate/20150106183347_create_votes_table.rb [hunk 6]
```diff
@@ -0,0 +1,11 @@
...
+      t.boolean :down_vote, null: false, default: false
+      t.integer :review_id
```
Comment (dpickett) [inline db/migrate/20150106183347_create_votes_table.rb]: can a vote exist without a review?

File: db/migrate/20150106183347_create_votes_table.rb [hunk 7]
```diff
@@ -0,0 +1,11 @@
...
+      t.integer :user_id
```
Comment (dpickett) [inline db/migrate/20150106183347_create_votes_table.rb]: can a vote exist without a user?

File: config/routes.rb [hunk 8]
```diff
@@ -1,6 +1,10 @@
...
+  end
+
+  resources :reviews, only: [:show] do
+    resources :votes, only: [:create, :update, :destroy]
```
Comment (dpickett) [inline config/routes.rb]: consider creating different controllers for upvotes and downvotes and posting to urls like:

```no-highlight
/reviews/23132/upvotes

/reviews/213214/downvotes
```

File: db/migrate/20150106195934_change_votes_table_to_one_vote_column.rb [hunk 9]
```diff
@@ -0,0 +1,7 @@
+class ChangeVotesTableToOneVoteColumn < ActiveRecord::Migration
+  def change
+    remove_column :votes, :up_vote
+    remove_column :votes, :down_vote
+    add_column :votes, :choice, :boolean, null: false
```
Comment (dpickett) [inline db/migrate/20150106195934_change_votes_table_to_one_vote_column.rb]: good work! Disregard last comments :-)

File: db/migrate/20150107050243_add_index_to_votes.rb [hunk 10]
```diff
@@ -0,0 +1,6 @@
+class AddIndexToVotes < ActiveRecord::Migration
+  def change
+    add_index :votes, [:user_id, :review_id], unique: true
+    add_index :votes, :review_id
```
Comment (dpickett) [inline db/migrate/20150107050243_add_index_to_votes.rb]: it's also likely that you want to index the `:user_id` - there is a use case for seeing a given user's votes.

File: app/controllers/votes_controller.rb (hunk 1 above)
Comment (nickcluc) [inline app/controllers/votes_controller.rb]: It was the only way we could get it to work :) ill try to refactor it.

**Pull request merged**